### ISSUE TITLE
Dependency Lock not failing the build when using older dependency

### ISSUE THREAD

**Issue by tsweets** [gradle/gradle#6130]
### Expected Behavior
If I have a dependency lock file that states Guava version 19 and my build uses version 17.0, then I expect the build to fail.

### Current Behavior
The build will pass if I use a version less than what is required or if I use latest ('+')
This should be a failure

### Context
Dependency Locking is not truly locking, thus this is not a a reproducible build. I'm also producing  a fat jar - it appears that this jar has the incorrect version as well. 

### Steps to Reproduce (for bugs)
Clone this project https://github.com/tsweets/dep-demo-bug.git
and run gradlew clean build
This should fail but it passes

### Your Environment
 * Build scan URL: https://gradle.com/s/ewzfpled3cmhy

**Comment by ljacomet** [comment]
Thanks for the report and the provided reproducible test case.

There are a couple changes to a build file that dependency locking does not fail upon. Downgrading a dependency is one of them.

However, downgrading a locked dependency has _no effect_ on locked configurations, the dependency is still using the locked version. So the changes that are _accepted_ are all changes that do not affect the resolution result and are thus _safe_.

It can be shown in your sample by running the following:

```
$ ./gradlew dependencyInsight --dependency guava --configuration compileClasspath

> Task :dependencyInsight
com.google.guava:guava:19.0
   variant "default" [
      org.gradle.status = release (not requested)
      Requested attributes not found in the selected variant:
         org.gradle.usage  = java-api
   ]
   Selection reasons:
      - Was requested
      - By constraint : dependency was locked to version '19.0'
      - By conflict resolution : between versions 17.0 and 19.0

com.google.guava:guava:19.0
\--- compileClasspath

com.google.guava:guava:17.0 -> 19.0
\--- compileClasspath

A web-based, searchable dependency report is available by adding the --scan option.

BUILD SUCCESSFUL in 0s
1 actionable task: 1 executed
```

However your `testRuntimeClasspath` configuration is not locked and thus there, `guava:17.0` is indeed used. If you locked that configuration with `guava:19.0`, you would see the expected output.

What you are hitting here is another issue: consistency of resolved dependencies across configurations. Locking does not solve that particular problem on its own.

**Comment by tsweets** [comment]
Thanks for the quick response.  I recreated the locks file with the other configurations and it did work as expected this time.  I'm not sure how I feel about the automatic upgrade to version 19 in the case where my dependency says 17. That might make it harder to debug an issue. However with that said our main goal is to ensure a certain set of dependencies are used. So this works great for that. 
Closing issue.
